pkg/storage/stores/indexshipper/uploads/index_set.go

```diff
@@ -0,0 +1,251 @@
+
+	for name, idx := range t.index {
+		// if the file is uploaded already do not upload it again.
+		t.indexUploadTimeMtx.RLock()
+		_, ok := t.indexUploadTime[name]
+		t.indexUploadTimeMtx.RUnlock()
+
+		if ok {
+			continue
+		}
+
+		if err := t.uploadIndex(ctx, idx); err != nil {
+			return err
+		}
+
+		t.indexUploadTimeMtx.Lock()
+		t.indexUploadTime[name] = time.Now()
+		t.indexUploadTimeMtx.Unlock()
+	}
+
+	level.Info(util_log.Logger).Log("msg", fmt.Sprintf("finished uploading table %s", t.tableName))
+
+	return nil
+}
+
+// Close Closes references to all the indexes.
+func (t *indexSet) Close() {
+	t.indexMtx.Lock()
+	defer t.indexMtx.Unlock()
+
+	for name, idx := range t.index {
+		if err := idx.Close(); err != nil {
+			level.Error(t.logger).Log("msg", fmt.Sprintf("failed to close index %s", name), "err", err)
+		}
+	}
+
+	t.index = map[string]index.Index{}
+}
+
+func (t *indexSet) uploadIndex(ctx context.Context, idx index.Index) error {
+	fileName := idx.Name()
+	level.Debug(t.logger).Log("msg", fmt.Sprintf("uploading index %s", fileName))
+
+	idxPath := idx.Path()
+
+	filePath := fmt.Sprintf("%s%s", idxPath, tempFileSuffix)
+	f, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := f.Close(); err != nil {
+			level.Error(util_log.Logger).Log("msg", "failed to close temp file", "path", filePath, "err", err)
+		}
+
+		if err := os.Remove(filePath); err != nil {
+			level.Error(util_log.Logger).Log("msg", "failed to remove temp file", "path", filePath, "err", err)
+		}
+	}()
+
+	compressedWriter := chunkenc.Gzip.GetWriter(f)
+	defer chunkenc.Gzip.PutWriter(compressedWriter)
+
+	idxReader, err := idx.Reader()
+	if err != nil {
+		return err
+	}
+
+	_, err = idxReader.Seek(0, 0)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(compressedWriter, idxReader)
+	if err != nil {
+		return err
+	}
+
+	err = compressedWriter.Close()
+	if err != nil {
+		return err
+	}
+
+	// flush the file to disk and seek the file to the beginning.
+	if err := f.Sync(); err != nil {
+		return err
+	}
+
+	if _, err := f.Seek(0, 0); err != nil {
+		return err
+	}
+
+	return t.storageIndexSet.PutFile(ctx, t.tableName, t.userID, t.buildFileName(fileName), f)
+}
+
+// Cleanup removes indexes which are already uploaded and have been retained for period longer than indexRetainPeriod since they were uploaded.
+func (t *indexSet) Cleanup(indexRetainPeriod time.Duration) error {
+	level.Info(util_log.Logger).Log("msg", fmt.Sprintf("cleaning up unwanted indexes from table %s", t.tableName))
+
+	var filesToCleanup []string
+	cutoffTime := time.Now().Add(-indexRetainPeriod)
+
+	t.indexMtx.RLock()
+
+	for name := range t.index {
+		t.indexMtx.RLock()
+		indexUploadTime, ok := t.indexUploadTime[name]
+		t.indexMtx.RUnlock()
+
+		if ok && indexUploadTime.Before(cutoffTime) {
+			filesToCleanup = append(filesToCleanup, name)
+		}
+	}
+
+	t.indexMtx.RUnlock()
```

Comment:
did you mean to wrap this loop in the `RLock` call as well as use `RLock` again on the same mutex within the loop? I would think you only meant to keep one of those?

Reply:
Sorry, my bad. I should be using `indexUploadTimeMtx`. This mistake showed us that you can take multiple read locks from the same goroutine which doesn't make a lot of sense.